### PR TITLE
Wrappers for the wallet's environment

### PR DESCRIPTION
**Pull request by LukasKorba**
Issue #197 

FYI, the code is slightly bigger that the issue associated, tasks done:
- TCA lightweight wallet wrapper for the keychain storage `WalletStorageInterator`
- SDK integrated, wallet's environment `WalletEnvironment` prepared as well as TCA's `WalletEnvironmentInterator`

Knowing the iInitialised wallet is a combination of keys present in the keychain + database files SDK is handling. We will use WalletEnvironmentInterator as the environment that can handle such functions. To create, validate and check presence of the wallet.

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: secantTests/Util/RecoveryPhraseStorageTests.swift
Comment (pacu) [inline secantTests/Util/RecoveryPhraseStorageTests.swift:41]: typo: testWalletStoredSucessfuly
testWalletStoredSuccessfully

File: secant/Util/WalletStorage.swift
Comment (pacu) [inline secant/Util/WalletStorage.swift:91]: complete TODO or add Follow-up github ticket for it

**Review by pacu — CHANGES_REQUESTED** [review]

File: secant/Util/WalletEnvironment.swift
Comment (LukasKorba) [reply secant/Util/WalletEnvironment.swift:64]: fixed

File: secantTests/Util/RecoveryPhraseStorageTests.swift
Comment (LukasKorba) [reply secantTests/Util/RecoveryPhraseStorageTests.swift:41]: fixed

File: secant/Util/WalletEnvironment.swift
Comment (LukasKorba) [reply secant/Util/WalletEnvironment.swift:13]: At this early stage of connecting things together I can't say it's a string that is going to be rendered to the user. I realised there is "some" failure and I felt like I want to at least explain what kind of failure it is. Probably to developers. Other option is to create specific failure case for each error. I'll remove it for now and leave `initializationFailed` only

File: secant/Util/WalletEnvironment.swift
Comment (LukasKorba) [reply secant/Util/WalletEnvironment.swift:22]: fixed

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:91]: fixed

File: secant/Util/WalletStorage.swift
Comment (adamstener) [inline secant/Util/WalletStorage.swift:24]: Does this need to be a class?

File: secant/Util/WalletStorage.swift
Comment (adamstener) [inline secant/Util/WalletStorage.swift:39]: Curious, why all the static functions?  Can we not use dependency injection?

**Review by adamstener — CHANGES_REQUESTED** [review]

**Review by pacu — APPROVED** [review]
LGTM!

File: secant/Util/WalletEnvironment.swift
Comment (pacu) [reply secant/Util/WalletEnvironment.swift:26]: Good catch! I think we need a TODO comment referring to issue #202

File: secant/Util/WalletEnvironment.swift
Comment (pacu) [reply secant/Util/WalletEnvironment.swift:17]: This is a really insightful information. It kind of leads to the conversation on how to integrate a synchronizer to the app in TCA's terms.

Months ago we agreed on the wallet being state driven, and that such state is defined by what the wallet finds on the blockchain when syncing with the keys provided by the user. Most screens need the synchronizer, but maybe not ALL OF IT, just a part of the state that it reflects. For example, Wallet details needs to get the Unified Address and decompose it in its receiver and nothing more, while the home screen needs the balance, sync progress and latest transactions, but does not need the UA, the send flow requires the synchronizer for getting the balance and the private keys to actually invoke `sendToAddress` to create and submit the transaction.

Comment (adamstener) [reply secant/Util/WalletEnvironment.swift:17]: I don't see a reason that any of these need to be classes, to be honest, it looks like everything in this PR can be structs but do let me know if I"m missing something.  I'm not sure what the synchronizer is either, and I'm not sure how it plays with the wallet information here, but it _sounds_ like that would just be another dependency to inject into stores, potentially.

Comment (adamstener) [reply secant/Util/WalletEnvironment.swift:17]: We don't need to use the interactor terminology, it was just something we added to Mnemonic because Mnemonic wasn't designed to have a public api like we wanted for TCA usage.

File: secant/Util/WalletEnvironment.swift
Comment (LukasKorba) [reply secant/Util/WalletEnvironment.swift:26]: removed completely, it'll be handled differently after issue #201 is merged and issue #220 implemented

File: secant/Util/WalletEnvironment.swift
Comment (LukasKorba) [reply secant/Util/WalletEnvironment.swift:18]: removed completely, it'll be handled differently after issue #201 is merged and issue #220 implemented

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:24]: struct from now on

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:188]: it doesn't, changed

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:39]: it doesn't, changed

File: secant/Util/WalletEnvironment.swift
Comment (LukasKorba) [reply secant/Util/WalletEnvironment.swift:17]: removed completely, will be handled by issue #220

File: secant/MockedDependencies/Services.swift
Comment (adamstener) [inline secant/MockedDependencies/Services.swift:13]: Should the name keyStorage be updated here now that we have a new type name?

File: secant/Util/WalletStorage.swift
Comment (adamstener) [inline secant/Util/WalletStorage.swift:12]: Do you think this is better suited to live inside the namespace of WalletStorage?

File: secantTests/Util/RecoveryPhraseStorageTests.swift
Comment (adamstener) [inline secantTests/Util/RecoveryPhraseStorageTests.swift:2]: Don't forget to update the file name

**Review by adamstener — CHANGES_REQUESTED** [review]

File: secant/Util/WalletStorageInteractor.swift
Comment (adamstener) [inline secant/Util/WalletStorageInteractor.swift:18]:
```suggestion
struct StoredWallet: Codable, Equatable {
    let language: MnemonicLanguageType
    let seedPhrase: String
    let version: Int
    
    var birthday: BlockHeight?
    var hasUserPassedPhraseBackupTest: Bool
}
```
Constants come before mutable properties according to our style guide 👍

File: secant/MockedDependencies/Services.swift
Comment (LukasKorba) [reply secant/MockedDependencies/Services.swift:13]: fixed

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:12]: fixed

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:183]: those should be 
```private extension WalletStorage {``` 
and then the separation makes sense, fixed

File: secantTests/Util/RecoveryPhraseStorageTests.swift
Comment (LukasKorba) [reply secantTests/Util/RecoveryPhraseStorageTests.swift:2]: fixed

File: secantTests/Util/RecoveryPhraseStorageTests.swift
Comment (LukasKorba) [reply secantTests/Util/RecoveryPhraseStorageTests.swift:29]: interactor removed completely

File: secantTests/Util/WalletStorageTests.swift
Comment (adamstener) [inline secantTests/Util/WalletStorageTests.swift:37]: I see we've removed the interactor but we're still relying on real data here. That makes these more or less integration tests.  Do you think it makes sense to wrap the Security dependency so that we can control it and make these unit tests?

File: secant/Util/WalletStorage.swift
Comment (pacu) [inline secant/Util/WalletStorage.swift:131]: I think a tab just sneaked in

File: secantTests/Util/WalletStorageTests.swift
Comment (pacu) [reply secantTests/Util/WalletStorageTests.swift:37]: Are you referring to mocking the underlying keychain calls?

**Review by pacu — COMMENTED** [review]
some comment about a tab sneaking in

File: secantTests/Util/WalletStorageTests.swift
Comment (adamstener) [reply secantTests/Util/WalletStorageTests.swift:37]: Then security framework has static functions we are calling here and relying on in tests.  I'm suggesting we wrap that functionality so we don't have to rely on the real data to write tests.

Comment (pacu) [reply secantTests/Util/WalletStorageTests.swift:37]: agreed. 

I don't think it's too much work to mock those two calls.

**Review by pacu — CHANGES_REQUESTED** [review]
LGTM.

I'd like to us to keep the integration tests and also have unit tests that don't rely in the real keychain.

File: secantTests/Util/WalletStorageTests.swift
Comment (LukasKorba) [reply secantTests/Util/WalletStorageTests.swift:37]: I must say I'm a bit confused... this PR is "only" naming refactor and protocol oriented code -> value oriented code change.. The previous RecoveryPhraseStorage (including tests) has been approved already... these tests are not new, not different. Can we merge this and iterate over tests as a separated ticket? 

I agree we need to test live keychain implementation because this is a bare bone of the wallet. If this component fails, nothing will work. If we want to mock keychain and fake it, I'm ok, let's do that but don't block this one now.

@pacu @adamstener

File: secant/Util/WalletStorage.swift
Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:131]: This comment is now inside the struct, separating some helper functions, therefore indented

Comment (LukasKorba) [reply secant/Util/WalletStorage.swift:131]: solved

**Review by pacu — APPROVED** [review]

File: secantTests/Util/WalletStorageTests.swift
Comment (pacu) [reply secantTests/Util/WalletStorageTests.swift:37]: Resolving and deferring to issue
https://github.com/zcash/secant-ios-wallet/issues/231

**Pull request merged**
